File: samples/bot-teams-authentication/csharp/Dialogs/LogoutDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;

namespace Microsoft.BotBuilderSamples
{
    /// <summary>
    /// A dialog that handles user logout.
    /// </summary>
    public class LogoutDialog : ComponentDialog
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="LogoutDialog"/> class.
        /// </summary>
        /// <param name="id">The dialog ID.</param>
        /// <param name="connectionName">The connection name for the OAuth provider.</param>
        public LogoutDialog(string id, string connectionName)
            : base(id)
        {
            ConnectionName = connectionName;
        }

        /// <summary>
        /// Gets the connection name for the OAuth provider.
        /// </summary>
        protected string ConnectionName { get; }

        /// <summary>
        /// Called when the dialog is started and pushed onto the dialog stack.
        /// </summary>
        /// <param name="innerDc">The dialog context for the current turn of conversation.</param>
        /// <param name="options">Optional, initial information to pass to the dialog.</param>
        /// <param name="cancellationToken">A cancellation token.</param>
        /// <returns>A task that represents the work queued to execute.</returns>
        protected override async Task<DialogTurnResult> OnBeginDialogAsync(DialogContext innerDc, object options, CancellationToken cancellationToken = default)
        {
            var result = await InterruptAsync(innerDc, cancellationToken);
            if (result != null)
            {
                return result;
            }

            return await base.OnBeginDialogAsync(innerDc, options, cancellationToken);
        }

        /// <summary>
        /// Called when the dialog is continued, where it is the active dialog and the user replies with a new activity.
        /// </summary>
        /// <param name="innerDc">The dialog context for the current turn of conversation.</param>
        /// <param name="cancellationToken">A cancellation token.</param>
        /// <returns>A task that represents the work queued to execute.</returns>
        protected override async Task<DialogTurnResult> OnContinueDialogAsync(DialogContext innerDc, CancellationToken cancellationToken = default)
        {
            var result = await InterruptAsync(innerDc, cancellationToken);
            if (result != null)
            {
                return result;
            }

            return await base.OnContinueDialogAsync(innerDc, cancellationToken);
        }

        /// <summary>
        /// Checks for the "logout" command and signs the user out if found.
        /// </summary>
        /// <param name="innerDc">The dialog context for the current turn of conversation.</param>
        /// <param name="cancellationToken">A cancellation token.</param>
        /// <returns>A task that represents the work queued to execute.</returns>
        private async Task<DialogTurnResult> InterruptAsync(DialogContext innerDc, CancellationToken cancellationToken = default)
        {
            if (innerDc.Context.Activity.Type == ActivityTypes.Message)
            {
                var text = innerDc.Context.Activity.Text.ToLowerInvariant();

                // Allow logout anywhere in the command
                if (text.Contains("logout"))
                {
                    // The UserTokenClient encapsulates the authentication processes.
                    var userTokenClient = innerDc.Context.TurnState.Get<UserTokenClient>();
                    await userTokenClient.SignOutUserAsync(innerDc.Context.Activity.From.Id, ConnectionName, innerDc.Context.Activity.ChannelId, cancellationToken).ConfigureAwait(false);

                    await innerDc.Context.SendActivityAsync(MessageFactory.Text("You have been signed out."), cancellationToken);
                    return await innerDc.CancelAllDialogsAsync(cancellationToken);
                }
            }

            return null;
        }
    }
}

File: samples/bot-teams-authentication/python/dialogs/logout_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.dialogs import DialogTurnResult, ComponentDialog, DialogContext
from botbuilder.core import BotFrameworkAdapter
from botbuilder.core.cloud_adapter_base import CloudAdapterBase
from botbuilder.schema import ActivityTypes
import logging

class LogoutDialog(ComponentDialog):
    def __init__(self, dialog_id: str, connection_name: str):
        # Initializes the LogoutDialog with a dialog ID and OAuth connection name.
        super(LogoutDialog, self).__init__(dialog_id)

        self.connection_name = connection_name

    async def on_begin_dialog(self, inner_dc: DialogContext, options: object) -> DialogTurnResult:
        # Intercepts the dialog at the beginning to check for logout command.
        result = await self._interrupt(inner_dc)
        if result:
            return result
        return await super().on_begin_dialog(inner_dc, options)

    async def on_continue_dialog(self, inner_dc: DialogContext) -> DialogTurnResult:
        # Intercepts the dialog while it is running to check for logout command.
        result = await self._interrupt(inner_dc)
        if result:
            return result
        return await super().on_continue_dialog(inner_dc)

    async def _interrupt(self, inner_dc: DialogContext):
        # Checks for the 'logout' command and signs the user out if detected.
        if inner_dc.context.activity.type == ActivityTypes.message:
            text = inner_dc.context.activity.text.lower()
            if text == "logout":
                # Use UserTokenClient for CloudAdapter
                adapter = inner_dc.context.adapter
                
                # Try BotFrameworkAdapter method first (for compatibility)
                if hasattr(adapter, 'sign_out_user'):
                    await adapter.sign_out_user(inner_dc.context, self.connection_name)
                else:
                    # CloudAdapter uses UserTokenClient
                    token_client = inner_dc.context.turn_state.get(CloudAdapterBase.USER_TOKEN_CLIENT_KEY)
                    if token_client:
                        try:
                            await token_client.sign_out_user(
                                inner_dc.context.activity.from_property.id,
                                self.connection_name,
                                inner_dc.context.activity.channel_id
                            )
                        except Exception as ex:
                            logging.error(f"Error signing out user: {ex}")
                    else:
                        logging.warning("UserTokenClient not found in turn state; cannot sign out.")
                
                await inner_dc.context.send_activity("You have been signed out.")
                return await inner_dc.cancel_all_dialogs()
